### ISSUE TITLE
UI Freezing

### ISSUE THREAD

**Issue by NomadicWarrior** [arkasas/OggDecoder#1]
I met project which have using this library.

```
func decodeLoops(packs: [SoundData], completion: @escaping ((Bool) -> Void)) {
    print("1111- Enter")
    for pack in packs {
      
//      let isInWriteTransaction = realm.isInWriteTransaction
//      if isInWriteTransaction == false {
//        realm.beginWrite()
//      }
      
      var timeInterval = Date().timeIntervalSince1970
      let decoder = OGGDecoder()

      Ambience.allCases.forEach { ambience in
        
        let fileName = (ambience.stringValue ?? "")
        
        let oggURL = pack.destinationURL
          .appendingPathComponent("Loops")
          .appendingPathComponent(fileName)
          .appendingPathExtension("ogg")
        let wavURL = pack.destinationURL
          .appendingPathComponent("Loops")
          .appendingPathComponent(fileName)
          .appendingPathExtension("wav")
        guard FileManager.default.fileExists(atPath: oggURL.path) else { return }
        
        let queue = DispatchQueue.global(qos: .background)
          queue.async {
            print("1111-Start")
            decoder.decode(oggURL, into: wavURL) { result in
              DispatchQueue.main.async {
                completion(true)
                print("1111-End")
              }
            }
          }
          
//        let sound = Sound()
//        sound.type = ambience.loopType ?? .loop1
//        sound.soundFileName = ambience.stringValue
//        sound.index = ambience.index
//        pack.package.sounds.append(sound)
        
        timeInterval += 1
      }
      
//      if isInWriteTransaction == false {
//        try? self.realm.commitWrite()
//      }
      
    }
  }
```

app UI is freezing. I can see that completion handler shortly returns to main thread. I tried with many ways but this task is always stacks the UI.

**Comment by arkasas** [comment]
I was able to reproduce the issue. The fix is already merged: https://github.com/arkasas/OggDecoder/pull/2.
Please confirm that issue is resolved.

**Comment by NomadicWarrior** [comment]
Thank you very much. I'm planing to do it tomorrow 24th of January. I'll come with update message

**Comment by NomadicWarrior** [comment]
> I was able to reproduce the issue. The fix is already merged: #2.
> Please confirm that issue is resolved.

Thank you very much. I'm planing to do it tomorrow 24th of January. I'll come with update message

**Comment by NomadicWarrior** [comment]
> I was able to reproduce the issue. The fix is already merged: #2.
> Please confirm that issue is resolved.

Looks like it's still blocks the UI. In my case I'm calling function "decoder.decode(oggURL, into: wavURL)" in custom background thread. like this "myCuncurentQueue.async { }. At the same time I'm opening Onboarding screen by asigning UIWindow.rootViewController = myOnboardingController. App doesn't appears onboarding until decode job finished. May be its because `(void)decode:(NSURL *)oggFile into:(NSURL *)outputFile completion:(void (^)(BOOL))completion {
    dispatch_async(dispatch_get_main_queue(), ^{
        BOOL result = [self decode:oggFile into:outputFile];
        dispatch_async(dispatch_get_main_queue(), ^(void){
            completion(result);
        });
    });`

decode function returns to main queue.

**Comment by NomadicWarrior** [comment]
> > I was able to reproduce the issue. The fix is already merged: #2.
> > Please confirm that issue is resolved.
> 
> Looks like it's still blocks the UI. In my case I'm calling function "decoder.decode(oggURL, into: wavURL)" in custom background thread. like this "myCuncurentQueue.async { }. At the same time I'm opening Onboarding screen by asigning UIWindow.rootViewController = myOnboardingController. App doesn't appears onboarding until decode job finished. May be its because `(void)decode:(NSURL *)oggFile into:(NSURL *)outputFile completion:(void (^)(BOOL))completion { dispatch_async(dispatch_get_main_queue(), ^{ BOOL result = [self decode:oggFile into:outputFile]; dispatch_async(dispatch_get_main_queue(), ^(void){ completion(result); }); });`
> 
> decode function returns to main queue.

Is it possible that u can return same thread in when framework decode() function job finished. just call completion?(bool) without returning to main thread?

**Comment by NomadicWarrior** [comment]
Cool, Thank you very much! I can see you added concurrent queue. Works as I expect. But firstly today it was working as previously I didn't notice any changed  maybe I incorrectly updated the package
